### PR TITLE
Implement Chrono Miner teleportation, v2

### PR DESCRIPTION
**Pull request by ghost**
This improved implementation doesn't touch on Common at all, by using `INotifyHarvesterAction`. It also features a customizable tick interval.

### CODE REVIEW

**Comment by GraionDilach** [comment]
Oh... nice.

What this is missing compared to RA2 are the warpaway animation and sounds. Could those be added to the trait - asin play an animation and a sound where the Miner disappears and another where it pops up.

The RA2 values were https://github.com/Phrohdoh/oramod-ra2/blob/master/extra-files/rules.ini#L540
https://github.com/Phrohdoh/oramod-ra2/blob/master/extra-files/rules.ini#L5994
https://github.com/Phrohdoh/oramod-ra2/blob/master/extra-files/sound.ini#L1645

**Comment by ghost** [comment]
I think it would be best to add these to a new `ChronoTeleportActivity` or something to replace `SimpleTeleport`, perhaps? The animations are shared with other chronoshifting units. I think the sound might be distinct but I can't remember.

File: OpenRA.Mods.RA2/OpenRA.Mods.RA2.csproj [hunk 1]
```diff
@@ -63,6 +63,7 @@
         <Command type="AfterBuild" command="cp ${TargetFile}.mdb ../mods/ra2" workingdir="${ProjectDir}" />
       </CustomCommands>
     </CustomCommands>
+    <WarningLevel>4</WarningLevel>
```
Comment (phrohdoh) [inline OpenRA.Mods.RA2/OpenRA.Mods.RA2.csproj]: Please do not commit unrelated changes.
(this line and all above)

File: OpenRA.Mods.RA2/Traits/ChronoDelivery.cs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+		CPos? dest = null;
+		Activity nextActivity = null;
+		int ticksTillCheck = 0;
+
+		public ChronoDelivery(Actor self, ChronoDeliveryInfo info)
+		{
+			Info = info;
+		}
+
+		#region ITick implementation
```
Comment (phrohdoh) [inline OpenRA.Mods.RA2/Traits/ChronoDelivery.cs]: No `#region`s please.

File: OpenRA.Mods.RA2/Traits/ChronoDelivery.cs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+		private void CheckTeleport(Actor self)
+		{
+			// We're already here; no need to interfere.
+			if (self.Location == dest.Value)
+			{
+				Reset();
+				return;
+			}
+
+			IPositionable pos = self.Trait<IPositionable>();
```
Comment (phrohdoh) [inline OpenRA.Mods.RA2/Traits/ChronoDelivery.cs]: `var`

File: OpenRA.Mods.RA2/Traits/ChronoDelivery.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+		public ChronoDelivery(Actor self, ChronoDeliveryInfo info)
+		{
+			Info = info;
+		}
+
+		#region ITick implementation
+		public void Tick(Actor self)
+		{
+			if (dest != null)
```
Comment (phrohdoh) [inline OpenRA.Mods.RA2/Traits/ChronoDelivery.cs]: You can instead do this:

``` csharp
{
if (dest == null)
    return;
...
}
```

**Comment by GraionDilach** [comment]
The sound is distinct, the animations were shared which was bad for modding.

**Comment by phrohdoh** [comment]
It doesn't matter what was unique, just require what you need in the ctor of `ChronoTeleportActivity`, but that can be left for another PR, which upstream may want for the `ctnk` and others.

**Comment by cjshmyr** [comment]
Wondering if you want to use `Teleport` instead of `SimpleTeleport`, which already supports sounds and picking best-possible cells (what did RA2 do when there's a harv already at the refinery when another wants to teleport?)

**Comment by GraionDilach** [comment]
In RA2, it didn't teleported until the docking cell was free.

**Comment by phrohdoh** [comment]
@cjshmyr If that is the only refinery the harv began driving towards the target refinery.

**Comment by ghost** [comment]
Updated with fixes.

**Comment by phrohdoh** [comment]
``` diff
diff --git a/mods/ra2/rules/allied-vehicles.yaml b/mods/ra2/rules/allied-vehicles.yaml
index 874fa13..28c4731 100644
--- a/mods/ra2/rules/allied-vehicles.yaml
+++ b/mods/ra2/rules/allied-vehicles.yaml
@@ -58,7 +58,7 @@ cmin:
        SearchFromOrderRadius: 12
        HarvestVoice: Move
        DeliverVoice: Move
-    ChronoDelivery:
+   ChronoDelivery:
    Health:
        HP: 1000
    Armor:
```

Yaml rules use tabs and no spaces for indentation.
Please always test your changes/features before submitting them.

**Comment by ghost** [comment]
Fixed.

File: OpenRA.Mods.RA2/Traits/ChronoDelivery.cs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+		public void MovementCancelled(Actor self)
+		{
+			Reset();
+		}
+
+		public void Harvested(Actor self, ResourceType resource) { }
+		public void Docked() { }
+		public void Undocked() { }
+
+		void CheckTeleport(Actor self)
```
Comment (phrohdoh) [inline OpenRA.Mods.RA2/Traits/ChronoDelivery.cs]: Could you rename this method to `TeleportIfPossible`?
I would expect a method whose name beings with `Check` to return `bool` and be pure.

File: OpenRA.Mods.RA2/Traits/ChronoDelivery.cs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+{
+	public class ChronoDeliveryInfo : ITraitInfo, Requires<HarvesterInfo>
+	{
+		[Desc("The number of ticks between each check to see if we can teleport home.")]
+		public readonly int CheckTeleportDelay = 10;
+
+		public virtual object Create(ActorInitializer init) { return new ChronoDelivery(init.Self, this); }
+	}
+
+	public class ChronoDelivery : INotifyHarvesterAction, ITick
```
Comment (phrohdoh) [inline OpenRA.Mods.RA2/Traits/ChronoDelivery.cs]: Please rename this (and therefor the Info) to `ChronoResourceDelivery` just to be a bit more verbose.

File: OpenRA.Mods.RA2/Traits/ChronoDelivery.cs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Activities;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.RA2.Traits
+{
+	public class ChronoDeliveryInfo : ITraitInfo, Requires<HarvesterInfo>
+	{
+		[Desc("The number of ticks between each check to see if we can teleport home.")]
```
Comment (phrohdoh) [inline OpenRA.Mods.RA2/Traits/ChronoDelivery.cs]: `... if we can teleport to an unreserved refinery.` please.

File: OpenRA.Mods.RA2/Traits/ChronoDelivery.cs [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+using System;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Activities;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.RA2.Traits
+{
+	public class ChronoDeliveryInfo : ITraitInfo, Requires<HarvesterInfo>
```
Comment (phrohdoh) [inline OpenRA.Mods.RA2/Traits/ChronoDelivery.cs]: Please add a `Desc` to the Info class aswell.
This should generally describe what the `ChronoResourceDelivery` trait does and why someone may want it.

File: OpenRA.Mods.RA2/Traits/ChronoDelivery.cs (hunk 7 above)
Comment (ghost) [inline OpenRA.Mods.RA2/Traits/ChronoDelivery.cs]: Left as "the refinery", since we don't check for free refineries in this class.

**Comment by ghost** [comment]
Updated with fixes

File: OpenRA.Mods.RA2/OpenRA.Mods.RA2.csproj (hunk 1 above)
Comment (GraionDilach) [inline OpenRA.Mods.RA2/OpenRA.Mods.RA2.csproj]: You end up better if you just revert this whole file and add the `Compile Include` line manually.

**Comment by ghost** [comment]
Now that I'm back to working on ORA, what do we need to merge this? Just sounds & animations?

**Comment by phrohdoh** [comment]
Yes please.

**Comment by ghost** [comment]
Ok. Did we ever come to any sort of consensus as to which trait should play them? Should it just be the chrono miner trait? Or do we want to ensure that it's a part of the `Activity` meaning we need a new activity?

**Comment by penev92** [comment]
It should be the Teleport activity. At a glance it seems like you're using `SimpleTeleport`, so probably that one.

**Comment by phrohdoh** [comment]
@scshunt Has this been updated with an animation?

**Comment by ghost** [comment]
@Phrohdoh I haven't pushed the changes yet because I need to rewrite them so as not to hardcode the values. There's also a minor style issue: the `effect@75alpha` palette is based on `ra`, not `effect`, and I'm hesitant to add `effect@50alpha` lest I cause more confusion than a Yuri clone.

**Comment by phrohdoh** [comment]
This eventually causes completely idle harvesters. They aren't looking for ore or a refn.

screen shot 2015-12-05 at 11 59 53 am

**Comment by penev92** [comment]
That shouldn't be related. AFAIK we still have that harvester issue (for all harvesters), and this code shouldn't be able to fix or break that.

**Comment by phrohdoh** [comment]
I've never seen them completely lose orders and sit idly.
This may not break or fix it, but it could expose it more by creating a situation for it to happen.
Possible debugging :)

**Comment by ghost** [comment]
Do you have a consistent reproduction?

**Comment by GraionDilach** [comment]
That sheer amount of miners will probably result with that behaviour on bleed as well.

**Comment by phrohdoh** [comment]
Uh, oops!
So I pushed and totally messed up all of the PRs somehow.
Luckily https://github.com/scshunt/OpenRA/commit/f717dcae76e85caa37b6c16a298ec3d29a91ceda still exists.

**Comment by phrohdoh** [comment]
I was able to repro the above issue on bleed, so it isn't manufactured here.

**Comment by Iamgoofball** [comment]
https://github.com/OpenRA/ra2/pull/1

**Comment by Mailaender** [comment]
Superseded by https://github.com/OpenRA/ra2/pull/156.
